File: node/src/domains/email_outbox.js
// email — outbound email behind a provider PORT, two dangerous properties proven, matching python/go:
// (1) EXACTLY-ONCE DISPATCH: idempotent on the Idempotency-Key via scopedKey + claimOnce (the atomic storeDo RMW)
// — two processes racing one key produce ONE recorded message, the loser is served the winner. The slot is SCOPED
// to the authenticated caller (a key is PRIVATE). A same-key retry with ANY different message (recipient/subject/
// body/template) is 409 — never a silent re-send, never a dropped Bcc. (2) HEADER SAFETY: every header-bound field
// rejects CR/LF + control/NEL/line-separator (addresses via validEmail, the rendered subject via validHeaderText,
// AFTER template rendering) so a subject (or a template value rendered into it) can never open a second header line;
// we REJECT (422), stricter than silent sanitizers. OWNER-scoped (owner = requireIdentity, never a body field):
// another caller's message is 404. Append-only; durable (a keyed send dedups after restart). Offline: the default
// backend RECORDS to the store (the record IS the outbox); a real provider is the dispatch swap-point (INTEROP.md).
import { intParam, nextId, problem, requireIdentity, sendJSON, storeGet, storePut, storeValues, testNow } from '../core/runtime.js';
import { digestHex, scopedKey } from '../parts/digest.js';
import { claimOnce } from '../parts/idempotent_claim.js';
import { paginate } from '../parts/paginate.js';
import { isWellFormed, makeWellFormed } from '../parts/well_formed.js';

const ROUTE = 'POST /email_outbox/messages';     // the dedup-slot discriminator (per-operation, owner-scoped slot)
const MAX_SUBJECT_BYTES = 998;            // RFC 5322 §2.1.1 line length — a hard protocol limit

// an operator LIMIT: a positive int within the 2^53-safe range (so go/node/python AGREE ×3 — env-knob overflow class).
function envLimit(name, def) {
  const raw = process.env[name];
  if (raw === undefined) return def;
  const v = Number(raw);
  return Number.isSafeInteger(v) && v >= 1 ? v : def;
}
const MAX_RECIPIENTS = envLimit('EMAIL_MAX_RECIPIENTS', 50);
const MAX_BODY_BYTES = envLimit('EMAIL_MAX_BODY_BYTES', 262144);

// THE TEMPLATE REGISTRY (policy, code-reviewed) — id -> {subject, html, text}; same data + same render ×3. NEVER empty.
const TEMPLATES = {
  verify_email: { subject: 'Verify your email address', html: '<p>Hi {{name}},</p><p>Confirm your address: {{link}}</p>', text: 'Hi {{name}},\nConfirm your address: {{link}}' },
  reset_password: { subject: 'Reset your password', html: '<p>Hi {{name}},</p><p>Reset your password: {{link}}</p>', text: 'Hi {{name}},\nReset your password: {{link}}' },
  notify: { subject: 'New message: {{title}}', html: '<p>{{body}}</p>', text: '{{body}}' },
};
const PLACEHOLDER = /\{\{([A-Za-z0-9_]+)\}\}/g; // ASCII-explicit -> ×3 parity with py/go

const LOCAL_CHARS = new Set("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.!#$%&'*+/=?^_`{|}~-");
const LABEL = /^[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?$/;

// validEmail — a strict boundary validator: a SUPERSET of the connector email_domain extractor + RFC 5321 caps + the
// WHATWG dot-atom charset. Surrounding whitespace is REJECTED (not trimmed — trimming differs ×3). Valid addresses
// are ASCII, so .length (UTF-16 units) == code points == octets for anything that could pass -> the decision is ×3.
function validEmail(s) {
  if (typeof s !== 'string' || s === '' || s.length > 254) return false;
  if ((s.match(/@/g) || []).length !== 1) return false;
  const at = s.indexOf('@');
  const local = s.slice(0, at);
  const domain = s.slice(at + 1);
  if (local === '' || domain === '' || local.length > 64 || domain.length > 255) return false;
  for (const c of local) if (!LOCAL_CHARS.has(c)) return false;
  const labels = domain.split('.');
  if (labels.length < 2) return false;
  for (const lbl of labels) if (!LABEL.test(lbl)) return false;
  return true;
}

// validHeaderText — the header-injection wall: reject CR/LF + the rest of C0, DEL, C1 (incl. NEL) and U+2028/U+2029.
function validHeaderText(s) {
  for (const ch of s) {
    const c = ch.codePointAt(0);
    if (c < 0x20 || c === 0x7F || (c >= 0x80 && c <= 0x9F) || c === 0x2028 || c === 0x2029) return false;
  }
  return true;
}

// render — scan the TEMPLATE for {{key}} (never iterate data -> deterministic ×3); a placeholder with no data value
// -> ok:false (a 422). Single-pass (a substituted value is not re-scanned). Object.hasOwn guards prototype keys.
function render(tpl, data) {
  let ok = true;
  const out = tpl.replace(PLACEHOLDER, (_m, key) => {
    if (!Object.hasOwn(data, key)) { ok = false; return ''; }
    return data[key];
  });
  return { out, ok };
}

const h = (s) => digestHex(s);                  // pre-hash one field to fixed colon-free hex
const hl = (xs) => digestHex(...xs.map(h));      // pre-hash each list element -> injective (the scoped_key idiom)

// bodyHashOf — the fingerprint over EVERY message-determining REQUEST field (digestHex joins with ':' and is NOT
// injective for free text, so each variable-length field is PRE-HASHED first). An added bcc / a changed data value
// all drift the hash -> a same-key reuse with any different message is 409.
function bodyHashOf(frm, to, cc, bcc, reply, subject, html, text, tid, data) {
  const dparts = Object.keys(data).sort().flatMap((k) => [h(k), h(data[k])]);
  return digestHex('from', h(frm), 'to', hl(to), 'cc', hl(cc), 'bcc', hl(bcc), 'reply', hl(reply),
    'subj', h(subject), 'html', h(html), 'text', h(text), 'tid', h(tid), 'data', digestHex(...dparts));
}

// dispatch — the offline fake backend: the stored record IS the sent message (record-to-store). A real backend
// transmits here; the INTEROP swap-point. Called ONLY on a fresh claim, so a retried/raced send never sends twice.
function dispatch(rec) {}

const publicView = (m) => ({ id: m.id, from: m.from, to: m.to, cc: m.cc, bcc: m.bcc, reply_to: m.reply_to, subject: m.subject, created_at: m.created_at });

// undefined -> []; an array of strings -> itself; anything else -> null (invalid type)
function asStringArray(v) {
  if (v === undefined) return [];
  if (Array.isArray(v) && v.every((x) => typeof x === 'string')) return v;
  return null;
}

export async function emailOutboxSend(req, res, params, body) {
  const owner = await requireIdentity(req, res); // authenticated mutation (no/invalid token -> 401), before any write
  if (owner === null) return;
  if (!body || typeof body !== 'object' || Array.isArray(body)) return problem(res, 422, 'invalid body');
  if (typeof body.from !== 'string' || !validEmail(body.from)) return problem(res, 422, 'from is not a valid email address');
  const to = asStringArray(body.to);
  const cc = asStringArray(body.cc);
  const bcc = asStringArray(body.bcc);
  const reply = asStringArray(body.reply_to);
  if (to === null || cc === null || bcc === null || reply === null) return problem(res, 422, 'invalid body');
  for (const group of [to, cc, bcc, reply]) for (const a of group) if (!validEmail(a)) return problem(res, 422, 'a recipient address is not valid');
  if (to.length === 0) return problem(res, 422, 'to must contain at least one recipient');
  if (to.length + cc.length + bcc.length > MAX_RECIPIENTS) return problem(res, 422, 'too many recipients');
  const seen = new Set();
  for (const group of [to, cc, bcc]) for (const a of group) {
    if (seen.has(a)) return problem(res, 422, 'a recipient address is duplicated across to, cc and bcc');
    seen.add(a);
  }
  const hasTpl = body.template !== undefined && body.template !== null;
  const hasRaw = body.subject !== undefined || body.html !== undefined || body.text !== undefined;
  if (hasTpl && hasRaw) return problem(res, 422, 'provide either a template or subject and body, not both');
  let subject;
  let html;
  let text;
  let bodyHash;
  let data = {};
  if (hasTpl) {
    const t = body.template;
    if (typeof t !== 'object' || Array.isArray(t) || typeof t.id !== 'string') return problem(res, 422, 'invalid body');
    if (t.data !== undefined) {
      if (typeof t.data !== 'object' || t.data === null || Array.isArray(t.data) || !Object.values(t.data).every((v) => typeof v === 'string')) {
        return problem(res, 422, 'invalid body');
      }
      data = t.data;
    }
    if (!Object.hasOwn(TEMPLATES, t.id)) return problem(res, 422, 'unknown template');
    // CONTAIN the data values BEFORE render + fingerprint (a lone surrogate cannot be UTF-8-hashed -> would throw).
    const cdata = {};
    for (const k of Object.keys(data)) cdata[k] = makeWellFormed(data[k]);
    const tpl = TEMPLATES[t.id];
    const r1 = render(tpl.subject, cdata);
    const r2 = render(tpl.html, cdata);
    const r3 = render(tpl.text, cdata);
    if (!r1.ok || !r2.ok || !r3.ok) return problem(res, 422, 'template variable not provided');
    subject = makeWellFormed(r1.out); html = makeWellFormed(r2.out); text = makeWellFormed(r3.out);
    bodyHash = bodyHashOf(body.from, to, cc, bcc, reply, '', '', '', t.id, cdata);
  } else {
    if (body.subject !== undefined && typeof body.subject !== 'string') return problem(res, 422, 'invalid body');
    if (body.html !== undefined && typeof body.html !== 'string') return problem(res, 422, 'invalid body');
    if (body.text !== undefined && typeof body.text !== 'string') return problem(res, 422, 'invalid body');
    if (typeof body.subject !== 'string' || (body.html === undefined && body.text === undefined)) {
      return problem(res, 422, 'a raw send needs a subject and at least one of html or text');
    }
    // CONTAIN BEFORE the fingerprint (a lone surrogate from a \u-escape cannot be UTF-8-hashed).
    subject = makeWellFormed(body.subject); html = makeWellFormed(body.html || ''); text = makeWellFormed(body.text || '');
    bodyHash = bodyHashOf(body.from, to, cc, bcc, reply, subject, html, text, '', {});
  }
  // RENDER-THEN-VALIDATE: header-safety + bounds on the CONTAINED, rendered output.
  if (!validHeaderText(subject)) return problem(res, 422, 'subject must not contain control characters or line breaks');
  if (Buffer.byteLength(subject, 'utf8') > MAX_SUBJECT_BYTES) return problem(res, 422, 'subject is too long');
  if (Buffer.byteLength(html, 'utf8') + Buffer.byteLength(text, 'utf8') > MAX_BODY_BYTES) return problem(res, 422, 'message body is too large');
  const created_at = testNow(req);
  const build = (eid) => ({ id: eid, owner, from: body.from, to, cc, bcc, reply_to: reply, subject, html, text, created_at, body_hash: bodyHash });
  const key = req.headers['idempotency-key'];
  if (key === undefined) { // no key -> no dedupe (opt-in)
    const eid = await nextId('email_outbox_message');
    const rec = build(eid);
    await storePut('email_outbox_messages', String(eid), rec);
    dispatch(rec);
    return sendJSON(res, 201, publicView(rec));
  }
  let nKeys = 0; // a single opaque token; node comma-joins duplicate headers, so count raw header lines and reject >1 (×3)
  for (let i = 0; i < req.rawHeaders.length; i += 2) if (req.rawHeaders[i].toLowerCase() === 'idempotency-key') nKeys += 1;
  if (nKeys > 1) return problem(res, 422, 'Idempotency-Key must be a single value');
  if (!isWellFormed(key)) return problem(res, 422, 'Idempotency-Key must be non-empty with no control characters');
  const scoped = scopedKey(ROUTE, owner, key);
  let prior = await storeGet('email_outbox_messages', scoped); // fast path: a settled key never mints
  if (prior === undefined) {
    const eid = await nextId('email_outbox_message'); // mint BEFORE the claim (a race loser's id is a harmless gap)
    const rec = build(eid);
    prior = await claimOnce('email_outbox_messages', scoped, rec);
    if (prior.id === eid) dispatch(prior); // I won the claim -> send ONCE
  }
  if (prior.owner !== owner) return problem(res, 409, 'Idempotency-Key is not owned by this caller');
  if (prior.body_hash !== bodyHash) return problem(res, 409, 'Idempotency-Key reused with a different message');
  sendJSON(res, 201, publicView(prior));
}

export async function emailOutboxList(req, res) {
  const owner = await requireIdentity(req, res);
  if (owner === null) return;
  const rows = (await storeValues('email_outbox_messages')).filter((m) => m.owner === owner).sort((a, b) => a.id - b.id).map(publicView);
  const q = new URL(req.url, 'http://localhost').searchParams;
  const { items, next, ok } = paginate(rows, q.get('cursor') || '', q.get('limit') || '');
  if (!ok) return problem(res, 422, 'invalid cursor or limit');
  sendJSON(res, 200, { results: items, next_cursor: next });
}

export async function emailOutboxGet(req, res, params) {
  const owner = await requireIdentity(req, res);
  if (owner === null) return;
  const id = intParam(params.message_id);
  if (id === null) return problem(res, 422, 'invalid message id');
  // unbounded-safe: a single-record lookup by id (returns at most one row); OWNER-scoped — not-yours == 404
  for (const m of await storeValues('email_outbox_messages')) {
    if (m.id === id && m.owner === owner) return sendJSON(res, 200, publicView(m));
  }
  return problem(res, 404, 'message not found');
}

File: go/internal/domains/email_outbox/email_outbox.go
// Package email — outbound email behind a provider PORT, two dangerous properties proven, matching python/node:
// (1) EXACTLY-ONCE DISPATCH: sending is idempotent on the Idempotency-Key — the claim is ONE atomic
// read-modify-write through (*KV).Do (ClaimOnce), so two processes racing one key produce ONE recorded message and
// the loser is served the winner. The slot is SCOPED to the authenticated caller (ScopedKey) — a key is PRIVATE.
// A same-key retry with ANY different message (recipient/subject/body/template) is 409 — never a silent re-send,
// never a dropped Bcc. (2) HEADER SAFETY: every header-bound field rejects CR/LF + control/NEL/line-separator
// (addresses via validEmail, the rendered subject via validHeaderText, AFTER template rendering) so a subject (or a
// template value rendered into it) can never open a second header line; we REJECT (422), stricter than silent
// sanitizers. OWNER-scoped (owner = RequireIdentity, never a body field): another caller's message is 404.
// Append-only; durable (a keyed send dedups after restart). Offline: the default backend RECORDS to the store (the
// record IS the outbox); a real provider is the emailDispatch swap-point (INTEROP.md). Every route auth'd.
package email_outbox

import (
	"net/http"
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"app/internal/core"
	"app/internal/parts/digest"
	"app/internal/parts/idempotent_claim"
	"app/internal/parts/paginate"
	"app/internal/parts/well_formed"
)

const (
	emailRoute            = "POST /email_outbox/messages" // the dedup-slot discriminator (per-operation, owner-scoped slot)
	emailMaxSubjectBytes  = 998                    // RFC 5322 §2.1.1 line length — a hard protocol limit
	emailLocalCharset     = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.!#$%&'*+/=?^_`{|}~-"
)

// an operator LIMIT: a positive int within the 2^53-safe range (so go/node/python AGREE ×3 — the env-knob overflow
// class); anything else falls back to the default. ×3-identical with python/node.
func emailEnvLimit(name string, def int) int {
	raw, ok := os.LookupEnv(name)
	if !ok {
		return def
	}
	v, err := strconv.Atoi(raw)
	if err != nil || v < 1 || int64(v) > (1<<53)-1 {
		return def
	}
	return v
}

var (
	emailMaxRecipients = emailEnvLimit("EMAIL_MAX_RECIPIENTS", 50)
	emailMaxBodyBytes  = emailEnvLimit("EMAIL_MAX_BODY_BYTES", 262144)
	emailMessages      = core.NewKV[string, emailMessage]("email_outbox_messages")
	emailLabel         = regexp.MustCompile(`^[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?$`)
	emailPlaceholder   = regexp.MustCompile(`\{\{([A-Za-z0-9_]+)\}\}`) // ASCII-explicit -> ×3 parity with py/node
)

// THE TEMPLATE REGISTRY (policy, code-reviewed) — id -> {subject, html, text}; same data + same render ×3. NEVER empty.
var emailTemplates = map[string]map[string]string{
	"verify_email":   {"subject": "Verify your email address", "html": "<p>Hi {{name}},</p><p>Confirm your address: {{link}}</p>", "text": "Hi {{name}},\nConfirm your address: {{link}}"},
	"reset_password": {"subject": "Reset your password", "html": "<p>Hi {{name}},</p><p>Reset your password: {{link}}</p>", "text": "Hi {{name}},\nReset your password: {{link}}"},
	"notify":         {"subject": "New message: {{title}}", "html": "<p>{{body}}</p>", "text": "{{body}}"},
}

type emailMessage struct {
	Id        int      `json:"id"`
	Owner     string   `json:"owner"`
	From      string   `json:"from"`
	To        []string `json:"to"`
	Cc        []string `json:"cc"`
	Bcc       []string `json:"bcc"`
	ReplyTo   []string `json:"reply_to"`
	Subject   string   `json:"subject"`
	Html      string   `json:"html"`
	Text      string   `json:"text"`
	CreatedAt int64    `json:"created_at"`
	BodyHash  string   `json:"body_hash"`
}

// validEmail — a strict boundary validator: a SUPERSET of the connector email_domain extractor + RFC 5321 length
// caps + the WHATWG dot-atom charset. Surrounding whitespace is REJECTED (not trimmed — trimming differs ×3). All
// valid addresses are ASCII, so len (bytes) == code points for any address that could pass — the decision is ×3.
func validEmail(s string) bool {
	if s == "" || len(s) > 254 || strings.Count(s, "@") != 1 {
		return false
	}
	at := strings.IndexByte(s, '@')
	local, domain := s[:at], s[at+1:]
	if local == "" || domain == "" || len(local) > 64 || len(domain) > 255 {
		return false
	}
	for _, c := range local {
		if !strings.ContainsRune(emailLocalCharset, c) {
			return false
		}
	}
	labels := strings.Split(domain, ".")
	if len(labels) < 2 {
		return false
	}
	for _, lbl := range labels {
		if !emailLabel.MatchString(lbl) {
			return false
		}
	}
	return true
}

// validHeaderText — the header-injection wall: reject CR/LF + the rest of C0, DEL, C1 (incl. NEL) and U+2028/U+2029.
func validHeaderText(s string) bool {
	for _, c := range s { // range iterates runes (code points) -> matches python's per-codepoint check
		if c < 0x20 || c == 0x7F || (c >= 0x80 && c <= 0x9F) || c == 0x2028 || c == 0x2029 {
			return false
		}
	}
	return true
}

// emailRender — scan the TEMPLATE for {{key}} (never iterate data -> deterministic ×3); a placeholder with no data
// value -> ok=false (a 422). Single-pass (a substituted value is not re-scanned).
func emailRender(tpl string, data map[string]string) (string, bool) {
	complete := true
	out := emailPlaceholder.ReplaceAllStringFunc(tpl, func(m string) string {
		key := m[2 : len(m)-2] // strip {{ }}
		v, ok := data[key]
		if !ok {
			complete = false
			return ""
		}
		return v
	})
	return out, complete
}

func emailH(s string) string { return digest.DigestHex(s) } // pre-hash one field to fixed colon-free hex

func emailHL(xs []string) string { // pre-hash each list element -> injective (the scoped_key idiom)
	parts := make([]any, len(xs))
	for i, x := range xs {
		parts[i] = emailH(x)
	}
	return digest.DigestHex(parts...)
}

// emailBodyHash — the fingerprint over EVERY message-determining REQUEST field (digest_hex joins with ':' and is NOT
// injective for free text, so each variable-length field is PRE-HASHED first). An added bcc / a changed data value
// all drift the hash -> a same-key reuse with any different message is 409.
func emailBodyHash(frm string, to, cc, bcc, reply []string, subject, html, text, tid string, data map[string]string) string {
	keys := make([]string, 0, len(data))
	for k := range data {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	dparts := make([]any, 0, len(keys)*2)
	for _, k := range keys {
		dparts = append(dparts, emailH(k), emailH(data[k]))
	}
	return digest.DigestHex("from", emailH(frm), "to", emailHL(to), "cc", emailHL(cc), "bcc", emailHL(bcc),
		"reply", emailHL(reply), "subj", emailH(subject), "html", emailH(html), "text", emailH(text),
		"tid", emailH(tid), "data", digest.DigestHex(dparts...))
}

// emailDispatch — the offline fake backend: the stored record IS the sent message (record-to-store). A real backend
// transmits here; the INTEROP swap-point. Called ONLY on a fresh claim, so a retried/raced send never sends twice.
func emailDispatch(m emailMessage) {}

func emailPublic(m emailMessage) map[string]any {
	return map[string]any{"id": m.Id, "from": m.From, "to": m.To, "cc": m.Cc, "bcc": m.Bcc,
		"reply_to": m.ReplyTo, "subject": m.Subject, "created_at": m.CreatedAt}
}

func emailNonEmpty(xs []string) []string { // JSON omits an absent list as nil; normalize to [] so the stored shape matches py/node
	if xs == nil {
		return []string{}
	}
	return xs
}

func EmailOutboxSend(w http.ResponseWriter, r *http.Request) {
	// PARSE -> AUTH -> SEMANTIC (×3 with python's body-parse-before-Depends): a typed decode rejects a numeric
	// recipient / a non-string subject / a non-string template value as 422; a no-token caller is 401.
	in, ok := core.DecodeJSON[struct {
		From     *string  `json:"from"`
		To       []string `json:"to"`
		Cc       []string `json:"cc"`
		Bcc      []string `json:"bcc"`
		ReplyTo  []string `json:"reply_to"`
		Subject  *string  `json:"subject"`
		Html     *string  `json:"html"`
		Text     *string  `json:"text"`
		Template *struct {
			Id   string            `json:"id"`
			Data map[string]string `json:"data"`
		} `json:"template"`
	}](w, r)
	if !ok {
		return
	}
	owner, ok := core.RequireIdentity(w, r)
	if !ok {
		return
	}
	if in.From == nil || !validEmail(*in.From) {
		core.WriteProblem(w, 422, "from is not a valid email address")
		return
	}
	for _, group := range [][]string{in.To, in.Cc, in.Bcc, in.ReplyTo} {
		for _, addr := range group {
			if !validEmail(addr) {
				core.WriteProblem(w, 422, "a recipient address is not valid")
				return
			}
		}
	}
	recipients := len(in.To) + len(in.Cc) + len(in.Bcc)
	if len(in.To) == 0 {
		core.WriteProblem(w, 422, "to must contain at least one recipient")
		return
	}
	if recipients > emailMaxRecipients {
		core.WriteProblem(w, 422, "too many recipients")
		return
	}
	seen := map[string]bool{}
	for _, group := range [][]string{in.To, in.Cc, in.Bcc} {
		for _, a := range group {
			if seen[a] {
				core.WriteProblem(w, 422, "a recipient address is duplicated across to, cc and bcc")
				return
			}
			seen[a] = true
		}
	}
	hasTpl := in.Template != nil
	hasRaw := in.Subject != nil || in.Html != nil || in.Text != nil
	if hasTpl && hasRaw {
		core.WriteProblem(w, 422, "provide either a template or subject and body, not both")
		return
	}
	var subject, html, text, bodyHash string
	data := map[string]string{}
	if hasTpl {
		tpl, known := emailTemplates[in.Template.Id]
		if !known {
			core.WriteProblem(w, 422, "unknown template")
			return
		}
		// CONTAIN the data values BEFORE render + fingerprint (identity in go — json already substitutes U+FFFD —
		// but kept parallel with py/node where a lone surrogate would otherwise break the UTF-8 hash).
		if in.Template.Data != nil {
			for k, v := range in.Template.Data {
				data[k] = well_formed.MakeWellFormed(v)
			}
		}
		var s1, s2, s3 bool
		subject, s1 = emailRender(tpl["subject"], data)
		html, s2 = emailRender(tpl["html"], data)
		text, s3 = emailRender(tpl["text"], data)
		if !s1 || !s2 || !s3 {
			core.WriteProblem(w, 422, "template variable not provided")
			return
		}
		subject, html, text = well_formed.MakeWellFormed(subject), well_formed.MakeWellFormed(html), well_formed.MakeWellFormed(text)
		bodyHash = emailBodyHash(*in.From, in.To, in.Cc, in.Bcc, in.ReplyTo, "", "", "", in.Template.Id, data)
	} else {
		if in.Subject == nil || (in.Html == nil && in.Text == nil) {
			core.WriteProblem(w, 422, "a raw send needs a subject and at least one of html or text")
			return
		}
		subject = well_formed.MakeWellFormed(*in.Subject)
		if in.Html != nil {
			html = well_formed.MakeWellFormed(*in.Html)
		}
		if in.Text != nil {
			text = well_formed.MakeWellFormed(*in.Text)
		}
		bodyHash = emailBodyHash(*in.From, in.To, in.Cc, in.Bcc, in.ReplyTo, subject, html, text, "", data)
	}
	// RENDER-THEN-VALIDATE: header-safety + bounds on the CONTAINED, rendered output.
	if !validHeaderText(subject) {
		core.WriteProblem(w, 422, "subject must not contain control characters or line breaks")
		return
	}
	if len(subject) > emailMaxSubjectBytes {
		core.WriteProblem(w, 422, "subject is too long")
		return
	}
	if len(html)+len(text) > emailMaxBodyBytes { // go strings are UTF-8 bytes -> octet count, matches py/node encode-len
		core.WriteProblem(w, 422, "message body is too large")
		return
	}
	createdAt := core.TestNow(r)
	build := func(eid int) emailMessage {
		return emailMessage{Id: eid, Owner: owner, From: *in.From, To: emailNonEmpty(in.To), Cc: emailNonEmpty(in.Cc),
			Bcc: emailNonEmpty(in.Bcc), ReplyTo: emailNonEmpty(in.ReplyTo), Subject: subject, Html: html, Text: text,
			CreatedAt: createdAt, BodyHash: bodyHash}
	}
	key, hasKey := r.Header["Idempotency-Key"]
	if !hasKey { // no key -> no dedupe (opt-in)
		eid := core.NextID("email_outbox_message")
		rec := build(eid)
		emailMessages.Set(strconv.Itoa(eid), rec)
		emailDispatch(rec)
		core.WriteJSON(w, 201, emailPublic(rec))
		return
	}
	if len(key) > 1 {
		core.WriteProblem(w, 422, "Idempotency-Key must be a single value")
		return
	}
	if !well_formed.IsWellFormed(key[0]) {
		core.WriteProblem(w, 422, "Idempotency-Key must be non-empty with no control characters")
		return
	}
	scoped := digest.ScopedKey(emailRoute, owner, key[0])
	prior, settled := emailMessages.Get(scoped)
	if !settled {
		eid := core.NextID("email_outbox_message") // mint BEFORE the claim (a race loser's id is a harmless gap)
		rec := build(eid)
		prior = idempotent_claim.ClaimOnce(emailMessages, scoped, rec)
		if prior.Id == eid { // I won the claim -> send ONCE
			emailDispatch(prior)
		}
	}
	if prior.Owner != owner {
		core.WriteProblem(w, 409, "Idempotency-Key is not owned by this caller")
		return
	}
	if prior.BodyHash != bodyHash {
		core.WriteProblem(w, 409, "Idempotency-Key reused with a different message")
		return
	}
	core.WriteJSON(w, 201, emailPublic(prior))
}

func EmailOutboxList(w http.ResponseWriter, r *http.Request) {
	owner, ok := core.RequireIdentity(w, r)
	if !ok {
		return
	}
	rows := []emailMessage{}
	for _, m := range emailMessages.All() {
		if m.Owner == owner { // OWNER-scoped: only the caller's own sends ever leave the store
			rows = append(rows, m)
		}
	}
	sort.Slice(rows, func(i, j int) bool { return rows[i].Id < rows[j].Id })
	views := make([]map[string]any, len(rows))
	for i, m := range rows {
		views[i] = emailPublic(m)
	}
	q := r.URL.Query()
	page, next, valid := paginate.Paginate(views, q.Get("cursor"), q.Get("limit"))
	if !valid {
		core.WriteProblem(w, 422, "invalid cursor or limit")
		return
	}
	var nc any
	if next != "" {
		nc = next
	}
	core.WriteJSON(w, 200, map[string]any{"results": page, "next_cursor": nc})
}

func EmailOutboxGet(w http.ResponseWriter, r *http.Request) {
	owner, ok := core.RequireIdentity(w, r)
	if !ok {
		return
	}
	id, err := strconv.Atoi(r.PathValue("message_id"))
	if err != nil {
		core.WriteProblem(w, 422, "invalid message id")
		return
	}
	// unbounded-safe: a single-record lookup by id (returns at most one row); OWNER-scoped — not-yours == 404
	for _, m := range emailMessages.All() {
		if m.Id == id && m.Owner == owner {
			core.WriteJSON(w, 200, emailPublic(m))
			return
		}
	}
	core.WriteProblem(w, 404, "message not found")
}
